File: atd/sync.py
"""git 同步：数据目录是 git 仓库，atd sync = commit + pull(并集合并) + push。

冲突处理：tasks.jsonl 一行一任务、行首是 id。若 rebase 冲突，把双方行按 id
做并集：同一 id 取 modified 新者；本地 tombstone（删除）优先于远端旧编辑。
"""
from __future__ import annotations

import json
import subprocess
import sys
from datetime import datetime
from pathlib import Path

from . import config
from .model import load_jsonl
from .storage import Store


def _git(dir_: Path, *args: str, check: bool = True) -> subprocess.CompletedProcess:
    r = subprocess.run(["git", "-C", str(dir_), *args],
                       capture_output=True, text=True, encoding="utf-8", errors="replace")
    if check and r.returncode != 0:
        raise RuntimeError(f"git {' '.join(args)} 失败：{r.stderr.strip() or r.stdout.strip()}")
    return r


def ensure_repo(dir_: Path | None = None) -> Path:
    dir_ = dir_ or config.data_dir()
    if not (dir_ / ".git").exists():
        _git(dir_, "init")
    # 数据目录里 undo/锁/临时文件不该进版本库
    gi = dir_ / ".gitignore"
    if not gi.exists():
        gi.write_text(".lock\nundo.jsonl\narchive.jsonl\n", encoding="utf-8")
    return dir_


def has_remote(dir_: Path) -> bool:
    r = _git(dir_, "remote", check=False)
    return bool(r.stdout.strip())


def _commit_all(dir_: Path, msg: str) -> bool:
    _git(dir_, "add", "-A")
    r = _git(dir_, "diff", "--cached", "--quiet", check=False)
    if r.returncode == 0:
        return False  # 没有变化
    _git(dir_, "commit", "-m", msg)
    return True


def _ours_id_map(text: str) -> dict[str, dict]:
    m: dict[str, dict] = {}
    for obj in load_jsonl(text):
        m[obj.get("id", "")] = obj
    return m


def _merge_union(ours_lines: list[str], theirs_lines: list[str]) -> list[str]:
    """按 id 并集合并两版 JSONL：新 modified 胜；tombstone 胜过旧编辑。"""
    ours = _ours_id_map("\n".join(ours_lines))
    theirs = _ours_id_map("\n".join(theirs_lines))
    out: dict[str, dict] = {}

    def mod_ts(o: dict) -> datetime:
        try:
            return datetime.fromisoformat(o.get("modified", ""))
        except ValueError:
            return datetime.min

    for tid, obj in ours.items():
        out[tid] = obj
    for tid, tobj in theirs.items():
        if tid not in out:
            out[tid] = tobj
            continue
        o = out[tid]
        if o.get("deleted") and not tobj.get("deleted"):
            continue  # 本地已删除，保留删除
        if tobj.get("deleted") and not o.get("deleted"):
            out[tid] = tobj  # 远端已删除 → 尊重删除
            continue
        if mod_ts(tobj) > mod_ts(o):
            out[tid] = tobj
    return [json.dumps(o, ensure_ascii=False) for o in out.values()]


def _resolve_conflict_file(path: Path) -> None:
    """把带冲突标记的 tasks.jsonl 按"本地版 vs 远端版"并集合并后写回。"""
    text = path.read_text(encoding="utf-8")
    ours: list[str] = []
    theirs: list[str] = []
    in_ours = True  # 冲突块外的公共行两边都有，归入 ours 即可
    for ln in text.splitlines():
        if ln.startswith("<<<<<<<"):
            in_ours = True
        elif ln.startswith("======="):
            in_ours = False
        elif ln.startswith(">>>>>>>"):
            in_ours = True
        else:
            (ours if in_ours else theirs).append(ln)
    merged = _merge_union(ours, theirs)
    path.write_text("\n".join(merged) + ("\n" if merged else ""), encoding="utf-8")


def sync(dir_: Path | None = None, *, can_push: bool = True) -> str:
    dir_ = ensure_repo(dir_)
    _commit_all(dir_, "atd: sync")
    if not has_remote(dir_):
        return "没有配置远程仓库：本地已 commit（git remote add origin <url> 后即可同步）"
    _git(dir_, "fetch", "--all")
    branch = _git(dir_, "rev-parse", "--abbrev-ref", "HEAD").stdout.strip() or "master"
    remote_ref = f"origin/{branch}"
    behind = _git(dir_, "rev-list", "--count", f"HEAD..{remote_ref}", check=False)
    if behind.returncode != 0:
        # 远程还没有这个分支（首次连接空仓库）
        if can_push:
            _git(dir_, "push", "-u", "origin", branch)
            return f"远程为空：已推送并建立 {branch} 分支"
        return f"远程没有 {branch} 分支"
    if behind.stdout.strip() == "0":
        if can_push:
            _git(dir_, "push", check=False)
            return "已推送（远程无新变更）"
        return "已是最新"
    # 远端有新提交：尝试 rebase；tasks.jsonl 冲突时用并集合并落盘后 continue
    r = _git(dir_, "rebase", remote_ref, check=False)
    if r.returncode != 0:
        merged_any = False
        status = _git(dir_, "status", "--porcelain").stdout
        for line in status.splitlines():
            p = line[3:].strip().strip('"')
            fp = dir_ / p
            if not fp.exists():
                continue
            if p == "tasks.jsonl":
                _resolve_conflict_file(fp)
                text = fp.read_text(encoding="utf-8")
                if any(ln.startswith(("<<<<<<<", "=======", ">>>>>>>")) for ln in text.splitlines()):
                    _git(dir_, "rebase", "--abort", check=False)
                    raise RuntimeError("tasks.jsonl 冲突合并失败，已回滚，请手动处理")
                _git(dir_, "add", p)
                merged_any = True
            else:
                _git(dir_, "checkout", "--ours", p, check=False)
                _git(dir_, "add", p, check=False)
        env = {**__import__("os").environ, "GIT_EDITOR": "true"}
        c = subprocess.run(["git", "-C", str(dir_), "rebase", "--continue"],
                           capture_output=True, text=True, encoding="utf-8", errors="replace", env=env)
        if c.returncode != 0:
            _git(dir_, "rebase", "--abort", check=False)
            raise RuntimeError("rebase continue 失败，已回滚：" + (c.stdout + c.stderr)[:300])
        if not merged_any:
            pass
    if can_push:
        p = _git(dir_, "push", check=False)
        if p.returncode != 0:
            _git(dir_, "pull", "--rebase", check=False)
            _git(dir_, "push", check=False)
    return "同步完成（远端新变更已合并）"


def status(dir_: Path | None = None) -> str:
    dir_ = ensure_repo(dir_)
    n_new, n_mod = 0, 0
    r = _git(dir_, "status", "--porcelain")
    for line in r.stdout.splitlines():
        if line.startswith("??") or line.startswith(" M") or line.startswith("M"):
            if "tasks.jsonl" in line:
                n_mod += 1
            else:
                n_new += 1
    remote = has_remote(dir_)
    return ("远程：" if remote else "无远程，") + f"待提交变更 {n_mod + n_new} 项"

File: src/sync/sync.ts
import { mkdir, readFile, stat, writeFile } from "node:fs/promises";
import { join } from "node:path";

import lockfile from "proper-lockfile";

import { DomainEventBus, type DomainEvents } from "../core/events.js";
import { atomicWrite, recoverArchiveTransaction } from "../storage/store.js";
import { pathsFor } from "../storage/paths.js";
import { gitOrThrow, gitRunner, type GitRunner } from "./git.js";
import { mergeConflictText } from "./merge.js";

const exists = async (path: string): Promise<boolean> => {
  try { await stat(path); return true; } catch { return false; }
};

export const ensureRepo = async (dir: string, runner: GitRunner = gitRunner): Promise<void> => {
  await mkdir(dir, { recursive: true });
  if (!(await exists(join(dir, ".git")))) await gitOrThrow(runner, ["init"], dir);
  // Keep local privacy rules in Git's per-repository exclude file so the
  // whitelist commit never creates an untracked .gitignore that can block rebase.
  const ignore = join(dir, ".git", "info", "exclude");
  await mkdir(join(dir, ".git", "info"), { recursive: true });
  const tasks = join(dir, "tasks.jsonl");
  if (!(await exists(tasks))) await writeFile(tasks, "", "utf8");
  const required = [".lock", ".archive.txn.json", "undo.jsonl", "archive.jsonl", "config.toml", "hooks/", ""]; 
  let current = "";
  if (await exists(ignore)) current = await readFile(ignore, "utf8");
  const lines = new Set(current.split(/\r?\n/));
  for (const line of required) if (line && !lines.has(line)) lines.add(line);
  const next = `${[...lines].filter(Boolean).join("\n")}\n`;
  if (next !== current) await writeFile(ignore, next, "utf8");
};

const gitIdentityError = (): Error => new Error(
  "Git 未配置提交身份。请先执行：\n  git config --global user.name \"Your Name\"\n  git config --global user.email \"you@example.com\"",
);

const ensureGitIdentity = async (dir: string, runner: GitRunner): Promise<void> => {
  const name = await runner.run(["config", "--get", "user.name"], { cwd: dir });
  const email = await runner.run(["config", "--get", "user.email"], { cwd: dir });
  const envName = process.env.GIT_AUTHOR_NAME || process.env.GIT_COMMITTER_NAME;
  const envEmail = process.env.GIT_AUTHOR_EMAIL || process.env.GIT_COMMITTER_EMAIL;
  if ((!name.stdout.trim() && !envName) || (!email.stdout.trim() && !envEmail)) throw gitIdentityError();
};

const sensitiveTracked = ["config.toml", "archive.jsonl", "undo.jsonl", ".archive.txn.json", "hooks"];
const assertSensitiveUntracked = async (dir: string, runner: GitRunner): Promise<void> => {
  const tracked = await gitOrThrow(runner, ["ls-files", "--", ...sensitiveTracked], dir);
  if (tracked.stdout.trim()) throw new Error(`为保护隐私，sync 拒绝提交已被 Git 跟踪的敏感路径：${tracked.stdout.trim()}\n请先备份并执行 git rm --cached <path>，再重试。`);
};

const commitAll = async (dir: string, runner: GitRunner): Promise<boolean> => {
  await assertSensitiveUntracked(dir, runner);
  const staged = await gitOrThrow(runner, ["diff", "--cached", "--name-only"], dir);
  const extraStaged = staged.stdout.split(/\r?\n/).map((value) => value.trim()).filter((file) => file && file !== "tasks.jsonl");
  if (extraStaged.length > 0) throw new Error(`sync 拒绝修改用户暂存区；请先提交或取消暂存非任务文件：${extraStaged.join(", ")}`);
  await gitOrThrow(runner, ["add", "--", "tasks.jsonl"], dir);
  const diff = await runner.run(["diff", "--cached", "--quiet"], { cwd: dir });
  if (diff.exitCode === 0) return false;
  await gitOrThrow(runner, ["commit", "-m", "atd: sync"], dir);
  return true;
};

const currentBranch = async (dir: string, runner: GitRunner): Promise<string> => {
  const result = await gitOrThrow(runner, ["rev-parse", "--abbrev-ref", "HEAD"], dir);
  return result.stdout.trim() || "master";
};

/** 刚 init 还没有任何提交时 rev-parse HEAD 会失败，看状态不该因此报错 */
const currentBranchTolerant = async (dir: string, runner: GitRunner): Promise<string> => {
  const symbolic = await runner.run(["symbolic-ref", "--short", "HEAD"], { cwd: dir });
  if (symbolic.exitCode === 0 && symbolic.stdout.trim()) return symbolic.stdout.trim();
  const head = await runner.run(["rev-parse", "--abbrev-ref", "HEAD"], { cwd: dir });
  return head.exitCode === 0 && head.stdout.trim() ? head.stdout.trim() : "master";
};

type RemoteTarget = { name: string; ref: string; hasUpstream: boolean };
const remoteTarget = async (dir: string, branch: string, runner: GitRunner): Promise<RemoteTarget | undefined> => {
  const upstream = await runner.run(["rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}"], { cwd: dir });
  if (upstream.exitCode === 0 && upstream.stdout.trim()) {
    const ref = upstream.stdout.trim();
    const slash = ref.indexOf("/");
    return { name: slash > 0 ? ref.slice(0, slash) : ref, ref, hasUpstream: true };
  }
  const remotes = (await runner.run(["remote"], { cwd: dir })).stdout.split(/\r?\n/).map((value) => value.trim()).filter(Boolean);
  const name = remotes.includes("origin") ? "origin" : remotes[0];
  return name ? { name, ref: `${name}/${branch}`, hasUpstream: false } : undefined;
};

const conflictFiles = async (dir: string, runner: GitRunner): Promise<string[]> => {
  const status = await gitOrThrow(runner, ["status", "--porcelain"], dir);
  return status.stdout.split(/\r?\n/).filter(Boolean).map((line) => line.slice(3).trim().replaceAll('"', ""));
};

const resolveRebaseConflicts = async (dir: string, runner: GitRunner): Promise<void> => {
  const files = await conflictFiles(dir, runner);
  if (files.length === 0) {
    await runner.run(["rebase", "--abort"], { cwd: dir });
    throw new Error("rebase 失败但没有可解析的冲突文件，已回滚");
  }
  for (const file of files) {
    if (file === "tasks.jsonl") {
      const path = join(dir, file);
      const merged = mergeConflictText(await readFile(path, "utf8"));
      await atomicWrite(path, merged.trimEnd() ? merged.trimEnd().split(/\r?\n/) : []);
    } else {
      const ours = await runner.run(["checkout", "--ours", "--", file], { cwd: dir });
      if (ours.exitCode !== 0) {
        await runner.run(["rebase", "--abort"], { cwd: dir });
        throw new Error(`无法保留本地冲突文件 ${file}，已回滚`);
      }
    }
    await gitOrThrow(runner, ["add", "--", file], dir);
  }
  const continued = await runner.run(["rebase", "--continue"], { cwd: dir, env: { ...process.env, GIT_EDITOR: "true" } });
  if (continued.exitCode !== 0) {
    await runner.run(["rebase", "--abort"], { cwd: dir });
    throw new Error(`rebase continue 失败，已回滚：${continued.stderr || continued.stdout}`);
  }
};

const rebaseOnto = async (dir: string, ref: string, runner: GitRunner): Promise<void> => {
  const result = await runner.run(["rebase", ref], { cwd: dir });
  if (result.exitCode !== 0) await resolveRebaseConflicts(dir, runner);
};

const pushWithRecovery = async (dir: string, target: RemoteTarget, branch: string, canPush: boolean, runner: GitRunner): Promise<void> => {
  if (!canPush) return;
  const pushed = await runner.run(target.hasUpstream ? ["push"] : ["push", "-u", target.name, branch], { cwd: dir });
  if (pushed.exitCode === 0) return;
  await gitOrThrow(runner, ["fetch", target.name], dir);
  const refreshed = await runner.run(["rev-parse", "--verify", target.ref], { cwd: dir });
  if (refreshed.exitCode !== 0) throw new Error(`push 失败且远端 ${target.ref} 不存在：${pushed.stderr || pushed.stdout}`);
  await rebaseOnto(dir, target.ref, runner);
  await gitOrThrow(runner, ["push", "-u", target.name, branch], dir);
};

export const syncDirectory = async (dir: string, canPush = true, runner: GitRunner = gitRunner, events = new DomainEventBus()): Promise<string> => {
  await ensureRepo(dir, runner);
  const pendingEvents: Array<() => void> = [];
  const emit = <K extends keyof DomainEvents>(event: K, payload: DomainEvents[K]): void => { pendingEvents.push(() => events.emit(event, payload)); };
  const release = await lockfile.lock(dir, { lockfilePath: join(dir, ".lock"), retries: { retries: 5, minTimeout: 20, maxTimeout: 200 } });
  try {
    await recoverArchiveTransaction(pathsFor(dir));
    await ensureGitIdentity(dir, runner);
    await commitAll(dir, runner);
    const branch = await currentBranch(dir, runner);
    const target = await remoteTarget(dir, branch, runner);
    if (!target) {
      const message = "没有配置远程仓库：本地已 commit（git remote add origin <url> 后即可同步）";
      emit("sync.completed", { summary: { changedTaskIds: [], message } });
      return message;
    }
    await gitOrThrow(runner, ["fetch", target.name], dir);
    const remoteExists = await runner.run(["rev-parse", "--verify", target.ref], { cwd: dir });
    if (remoteExists.exitCode !== 0) {
      await pushWithRecovery(dir, target, branch, canPush, runner);
      const message = canPush ? `远程为空：已推送并建立 ${branch} 分支` : `远程没有 ${branch} 分支`;
      emit("sync.completed", { summary: { changedTaskIds: [], message } });
      return message;
    }
    const behind = await gitOrThrow(runner, ["rev-list", "--count", `HEAD..${target.ref}`], dir);
    if (behind.stdout.trim() !== "0") await rebaseOnto(dir, target.ref, runner);
    await pushWithRecovery(dir, target, branch, canPush, runner);
    const message = "同步完成（远端新变更已合并）";
    emit("sync.completed", { summary: { changedTaskIds: [], message } });
    return message;
  } catch (error) {
    const normalized = error instanceof Error ? error : new Error(String(error));
    emit("sync.failed", { error: normalized });
    throw normalized;
  } finally {
    await release();
    for (const publish of pendingEvents) publish();
  }
};

/**
 * 配置（或改写）origin。这一步之前只能自己敲 git remote add，
 * README 也只写了「git remote add origin <url>」，对不熟 git 的人是道坎。
 */
export const setupRemote = async (dir: string, url: string, runner: GitRunner = gitRunner): Promise<string> => {
  if (!url.trim()) throw new Error("远程地址不能为空");
  await ensureRepo(dir, runner);
  const remotes = (await runner.run(["remote"], { cwd: dir })).stdout.split(/\r?\n/).map((value) => value.trim()).filter(Boolean);
  if (remotes.includes("origin")) {
    const previous = (await runner.run(["remote", "get-url", "origin"], { cwd: dir })).stdout.trim();
    if (previous === url) return `origin 已经是 ${url}，没有改动`;
    await gitOrThrow(runner, ["remote", "set-url", "origin", url], dir);
    return `已把 origin 从 ${previous} 改成 ${url}`;
  }
  await gitOrThrow(runner, ["remote", "add", "origin", url], dir);
  return `已添加远程 origin → ${url}\n接下来跑一次 atd sync 就会把本地任务推上去`;
};

export type SyncStatusDetail = {
  branch: string;
  remote?: string;
  remoteUrl?: string;
  pending: number;
  ahead: number;
  behind: number;
  lastCommit?: string;
};

export const syncStatusDetail = async (dir: string, runner: GitRunner = gitRunner): Promise<SyncStatusDetail> => {
  await ensureRepo(dir, runner);
  const status = await gitOrThrow(runner, ["status", "--porcelain"], dir);
  const branch = await currentBranchTolerant(dir, runner);
  const target = await remoteTarget(dir, branch, runner);
  const detail: SyncStatusDetail = { branch, pending: status.stdout.split(/\r?\n/).filter(Boolean).length, ahead: 0, behind: 0 };
  const lastCommit = await runner.run(["log", "-1", "--format=%cd %s", "--date=format:%Y-%m-%d %H:%M"], { cwd: dir });
  if (lastCommit.exitCode === 0 && lastCommit.stdout.trim()) detail.lastCommit = lastCommit.stdout.trim();
  if (!target) return detail;
  detail.remote = target.name;
  const url = await runner.run(["remote", "get-url", target.name], { cwd: dir });
  if (url.exitCode === 0 && url.stdout.trim()) detail.remoteUrl = url.stdout.trim();
  // 只数已抓到的引用，不联网：status 要能在离线时立刻返回
  const counts = await runner.run(["rev-list", "--left-right", "--count", `${target.ref}...HEAD`], { cwd: dir });
  if (counts.exitCode === 0) {
    const [behind, ahead] = counts.stdout.trim().split(/\s+/u).map(Number);
    detail.behind = behind ?? 0;
    detail.ahead = ahead ?? 0;
  }
  return detail;
};

export const syncStatus = async (dir: string, runner: GitRunner = gitRunner): Promise<string> => {
  const detail = await syncStatusDetail(dir, runner);
  const lines = [
    `分支：${detail.branch}`,
    detail.remoteUrl ? `远程：${detail.remote} → ${detail.remoteUrl}` : "远程：未配置（atd sync --setup <url> 可以配）",
    `未提交变更：${detail.pending} 项`,
  ];
  if (detail.remote) lines.push(`领先远端 ${detail.ahead} 个提交，落后 ${detail.behind} 个（数字基于上次 fetch，未联网刷新）`);
  if (detail.lastCommit) lines.push(`最近一次提交：${detail.lastCommit}`);
  return lines.join("\n");
};
